controlplane/src/core/util.ts

```diff
@@ -371,3 +373,46 @@ export function getValueOrDefault<K, V>(map: Map<K, V>, key: K, constructor: ()
 export function webhookAxiosRetryCond(err: AxiosError) {
   return isNetworkError(err) || isRetryableError(err);
 }
+
+export function createS3ClientConfig(bucketName: string, opts: S3StorageOptions): S3ClientConfig {
+  const url = new URL(opts.url);
+  const { region, username, password } = opts;
+  const forcePathStyle = !isVirtualHostStyleUrl(url);
+  const endpoint = opts.endpoint || (forcePathStyle ? url.origin : url.origin.replace(`${bucketName}.`, ''));
+
+  const accessKeyId = url.username || username || '';
+  const secretAccessKey = url.password || password || '';
+
+  if (!accessKeyId || !secretAccessKey) {
+    throw new Error('Missing S3 credentials. Please provide access key ID and secret access key.');
+  }
+
+  if (!region) {
+    throw new Error('Missing region in S3 configuration.');
+  }
+
+  return {
+    region,
+    endpoint,
+    credentials: {
+      accessKeyId,
+      secretAccessKey,
+    },
+    forcePathStyle,
+  };
+}
+
+export function extractS3BucketName(s3Url: string) {
+  const url = new URL(s3Url);
+
+  if (isVirtualHostStyleUrl(url)) {
+    return url.hostname.split('.')[0];
+  }
+
+  // path based style
+  return url.pathname.slice(1);
+}
+
+export function isVirtualHostStyleUrl(url: URL) {
```

Comment:
Should we put these into `shared` package?

Reply:
Ah yes we should, thanks for the hint, will move it.